GAE/test/org/akvo/flow/api/app/DataPointServletTest.java

```diff
@@ -0,0 +1,214 @@
+        assertFalse(surveyAssignments.isEmpty());
+
+        final SurveyAssignment sa = surveyAssignments.get(0);
+        assertEquals(surveyId, sa.getSurveyId());
+        assertTrue(sa.getDeviceIds().contains(2L));
+        assertEquals(formIds, sa.getFormIds());
+    }
+
+    @Test
+    public void someDataPointsTest() {
+
+        final Long surveyId = 100L;
+        final List<SurveyedLocale> dataPoints = createDataPoints(surveyId, 10);
+
+        final SurveyedLocale dataPoint = dataPoints.get(7);
+        assertEquals("7", dataPoint.getIdentifier());
+
+        final List<Long> selectedDataPointIds = Arrays.asList(dataPoints.get(5).getKey().getId(), dataPoints.get(6).getKey().getId());
+        final Long assignmentId = 20L;
+        final Long deviceId = 1L;
+
+        createDataPointAssignment(assignmentId, deviceId, selectedDataPointIds, surveyId);
+
+        final DataPointServlet servlet = new DataPointServlet();
+
+        final List<SurveyedLocale> someDataPoints = servlet.getDataPointList(surveyId, deviceId);
+
+        final Set<Long> selectedIds = new HashSet<>(selectedDataPointIds);
+        final Set<Long> foundIds = getEntityIds(someDataPoints);
+
+        assertEquals(selectedIds, foundIds);
+    }
+
+
+    /**
+     * New schema: `[0]` as data point list represent all data points
+     **/
+    @Test
+    public void allDataPointsNewSchemaTest() {
+
+        final Long surveyId = 200L;
+        final List<SurveyedLocale> allDataPoints = createDataPoints(surveyId, 15);
+        final Long assignmentId = 30L;
+        final Long deviceId = 2L;
+
+        createDataPointAssignment(assignmentId, deviceId, Collections.singletonList(0L), surveyId);
+
+        final DataPointServlet servlet = new DataPointServlet();
+        final List<SurveyedLocale> foundDataPoints = servlet.getDataPointList(surveyId, deviceId);
+
+        final Set<Long> allDataPointIds = getEntityIds(allDataPoints);
+        final Set<Long> foundDataPointIds = getEntityIds(foundDataPoints);
+
+        assertEquals(allDataPointIds, foundDataPointIds);
+    }
+
+
+    /**
+     * Old schema: We have a `SurveyAssingment` but no `DataPointAssignment`
+     */
+    @Test
+    public void allDataPointsOldSchemaTest() {
+        final Long surveyId = 300L;
+        final List<SurveyedLocale> allDataPoints = createDataPoints(surveyId, 20);
+        final List<Long> deviceIds = Arrays.asList(5L, 6L);
+        final List<Long> formIds = Arrays.asList(10L, 11L);
+
+        final SurveyAssignment sa = createAssignment(surveyId, deviceIds, formIds);
+
+        final DataPointServlet servlet = new DataPointServlet();
+        final List<SurveyedLocale> foundDataPoints = servlet.getDataPointList(surveyId, 6L);
+
+        final Set<Long> allDataPointIds = getEntityIds(allDataPoints);
+        final Set<Long> foundDataPointIds = getEntityIds(foundDataPoints);
+
+        assertEquals(allDataPointIds, foundDataPointIds);
```

Comment:
This is duplicated several times.

Reply:
_"Duplicated code is better than a wrong abstraction"_ , we only have 3?

Reply:
_""Duplication is the root of all evil"_. Even with 1, it is going to make the tests more readable.